board_similarity.py

```
@@ -0,0 +1,161 @@
+                    x += 1
+                    result[-1].append((x, y))
+                else:
+                    x += 1
+
+    return result
+
+
+def pieces_similarity(fen1, fen2):
+    """
+    :param fen1: first fen
+    :param fen2: second fen
+    :return: similarity of pieces for given fens in the range from -1 to 1
+    """
+    coordinates1 = fen_to_coord(fen1)
+    coordinates2 = fen_to_coord(fen2)
+    result = 0
+    length = 0
+    for n_piece in range(12):
+        piece_result = 0
+        for coord in coordinates1[n_piece]:
+            if coord in coordinates2[n_piece]:
+                piece_result += 1
+        length += len(coordinates1[n_piece]) + len(coordinates2[n_piece]) - piece_result
+        result += piece_result
+    return result / length * 2 - 1
+
+
+def predict_similarity(
+    fen1,
+    fen2,
+    engine,
+    moves_deep=5,
+    time_limit=0.1,
+    moves_similarity_weight=1,
+    score_similarity_weight=1,
+    pieces_similarity_weight=1,
+    info1=None,
+    info2=None,
+):
+    """
+    :param fen1: first fen
+    :param fen2: second fen
+    :param engine: chess engine, in class chess.engine.SimpleEngine
+    :param depth: number of moves to be analyzed in the move_similarity function
+    :param time_limit: float of max time to analyze fen
+    :param moves_similarity_weight,score_similarity_weight,pieces_similarity_weight:
+    the weight of the result of a given similarity function
+    :return: similarity given fens in the range from -1 to 1
+    """
+    board1 = chess.Board(fen1)
+    board2 = chess.Board(fen2)
+    if info1 == None:
+        info1 = engine.analyse(
+            board1, chess.engine.Limit(depth=moves_deep, time=time_limit)
+        )
+    if info2 == None:
+        info2 = engine.analyse(
+            board2, chess.engine.Limit(depth=moves_deep, time=time_limit)
+        )
+    try :
+        ms = moves_similarity(info1["pv"], info2["pv"], board1, board2)
+    except:

```

It is a bad idea to catch all exceptions, because you may accidentally silence an exception you don't want to, e.g. one that means that moves_similarity is 100% broken